### DISCUSSION TITLE
Masking documentation

### DISCUSSION THREAD

**Discussion by dprevost-LMI** [webdriverio/webdriverio#13912] [Technical Discussion]
We have been using web driver IO for a while now, and we have implemented different strategies to mask our test user's password from different logs since we run in a public URL mainly to integrate with BrowserStack.

I just did a new search on the Wdio website and could not find a single page on our to mask any sensitive data. Even though most of the accounts and data are testing data only, having a masking strategy is crucial when dealing with auth tokens and passwords of keys so that no one can play around with them.

So why is there no documentation at all on masking?

Different touch points I'm wondering about:
- Appium log
- Allure report
- The send command
- Network logs

To share our current strategy:
When doing tests locally:
- For the appium log, I use the [logFilter](https://appium.io/docs/en/2.0/guides/log-filters/), and I change all passwords to contain a keyword like `@mask@` so I can have an easy regex. Warning: the password also appears with all the characters separated by a comma. Remember that case!
- When sending a password through `setValue`, we momentarily set the log level to error instead of info or debug using the `@wdio/logger` lib
- For the allure report, we have patched the lib so that when a command contains `@maks@`, I change the value for a masked one.

On BrowserStack:
- We had to disable network logs (`networkLogs: false`) not to leak authentication tokens
- For appium logs, we had to disable it too with `appiumLogs: false` since Appium log filter is not supported
- We have also used `maskCommands: 'setValues, getValues, setCookies, getCookies'` as [documented by BrowserStack](https://www.browserstack.com/docs/automate/selenium/hide-sensitive-data) to hide sensitive data.

So, let me know what you think and how you do the masking? If I get some good replies I might end up entering a bug in WDIO for better documentation

**Comment by christian-bromann** [comment]
> So why is there no documentation at all on masking?

Because there is no feature that allows masking in WebdriverIO logs. Feel free to raise a feature request and get involved helping to land this type of feature. I think it makes absolutely sense to have it.

**Comment by dprevost-LMI** [comment]
Great, I will raise a feature request and get involved on it.

I'm currently trying some changes, and I found that one of the most "secure" places to hook and mask text would be right before the `onBeforeCommand` is called. That would mask values from Reporters but also in BrowserStack, which I need.

In short, I would add the following to the [BrowserStack service.ts](https://github.com/dprevost-LMI/webdriverio/blob/abc2ec893b7597ac73b061f8617aeb7d90f88574/packages/wdio-browserstack-service/src/service.ts#L177-L188) and in the [reporter index.js](https://github.com/dprevost-LMI/webdriverio/blob/abc2ec893b7597ac73b061f8617aeb7d90f88574/packages/wdio-runner/src/index.ts#L285-L288)

```
const patternWithFlags = this._config.onBeforeCommandTextPatternMasker
const replacer = '**SECURE**'

let maskedCommand = command
if (!!patternWithFlags && !!command?.body?.text && command.body.text.match(patternWithFlags)) {
    maskedCommand = {
        ...command,
        body: {
            ...command.body,
            text: replacer
        }
    }
}
```

Why use a simple regex and replace the text value with a 'MASKED' value? This aligns with the [appium logFilter](https://appium.io/docs/en/2.0/guides/log-filters/) strategy so we can have a fully working solution!
 - Disclaimer: I wouldn't say I like the appium approach, but I could not find a better way.

By the way, if a feature request needs to be raised before answering, I will do it!

**Comment by dprevost-LMI** [comment]
A feature request was raised with a PR providing a first draft solution. See https://github.com/webdriverio/webdriverio/issues/13937

**Reply by christian-bromann** [reply]
Thanks @dprevost-LMI , once the PR is out of draft and passes tests I am happy to take a look!

**Reply by dprevost-LMI** [reply]
Great! I was waiting on a confirmation of the solution before doing everything around it, so if your comment above is that confirmation, I will unit-test it and document it better!
